contracts/airdrop.sol

```diff
@@ -0,0 +1,102 @@
+pragma solidity ^0.5.15;
+pragma experimental ABIEncoderV2;
+
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import {IERC20} from "./interfaces/IERC20.sol";
+import {ITokenRegistry} from "./interfaces/ITokenRegistry.sol";
+
+import {Types} from "./libs/Types.sol";
+import {RollupUtils} from "./libs/RollupUtils.sol";
+import {ECVerify} from "./libs/ECVerify.sol";
+
+import {MerkleTreeUtils as MTUtils} from "./MerkleTreeUtils.sol";
+
+contract AirdropSetup {
+    using SafeMath for uint256;
+    using ECVerify for bytes32;
+
+    MTUtils public merkleUtils;
+
+    bytes32
+        public constant ZERO_BYTES32 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+
+    /********************
+     * Error Codes *
+     ********************/
+    uint256 public constant NO_ERR = 0;
+    uint256 public constant ERR_TOKEN_ADDR_INVAILD = 1; // account doesnt hold token type in the tx
+    uint256 public constant ERR_TOKEN_AMT_INVAILD = 2; // tx amount is less than zero
+    uint256 public constant ERR_TOKEN_NOT_ENOUGH_BAL = 3; // leaf doesnt has enough balance
+    uint256 public constant ERR_FROM_TOKEN_TYPE = 4; // from account doesnt hold the token type in the tx
+    uint256 public constant ERR_TO_TOKEN_TYPE = 5; // to account doesnt hold the token type in the tx
+}
+
+contract AirdropHelpers is AirdropSetup {
+    function AddTokensToAccount(
+        Types.UserAccount memory account,
+        uint256 numOfTokens
+    ) public pure returns (Types.UserAccount memory updatedAccount) {
+        return (
+            RollupUtils.UpdateBalanceInAccount(
+                account,
+                RollupUtils.BalanceFromAccount(account).add(numOfTokens)
+            )
+        );
+    }
+    function UpdateAccountWithSiblings(
+        Types.UserAccount memory new_account,
+        Types.AccountMerkleProof memory _merkle_proof
+    ) public view returns (bytes32, uint256) {
+        bytes32 newRoot = merkleUtils.updateLeafWithSiblings(
+            keccak256(RollupUtils.BytesFromAccount(new_account)),
+            _merkle_proof.accountIP.pathToAccount,
+            _merkle_proof.siblings
+        );
+        uint256 balance = RollupUtils.BalanceFromAccount(new_account);
+        return (newRoot, balance);
+    }
+}
+
+contract Airdrop is AirdropHelpers {
+
+    function processDrop(
+        Types.Transaction memory drop,
+        Types.AccountMerkleProof memory _to_merkle_proof
+    )
+        public
+        view
+        returns (
+            bytes32,
+            uint256,
+            bool
+        )
+    {
```

Comment:
TODO: I think we still need to check the public key is available in the PDA tree.

Reply:
Yeah I think so too.

Reply:
Maybe add a TODO comment for signature check as well.

Reply:
For airdrop, we might not want to check the signature for each drop.

Reply:
why not?

Reply:
Ah, I mean we want to check 1 signature at the batch level, but don't want each transaction in that batch to have their own signature and check them many times. Does that make sense?

Reply:
I think we need to have one signature per tx. Because reddit will just sign a single message per user and then the user gives it to the coordinator.